### ISSUE TITLE
WebTrigger (WebAction async)

### ISSUE THREAD

**Issue by csantanapr** [apache/incubator-openwhisk#3558]
So far in multiple occasions have talk to customers/users using OpenWhisk that they have a need to call a Web Action but want the Action to return right away but continue to do business logic.

One tangible use case is Web Analytics:
Web sites will call a web url on APIGateway, they would get back 202 response right away with the activation id of the web action.
The action will process the http request, taking more time (i.e. 1 second). The web site doesn't have to wait 1 second for it.

The current way of achieving is not optimal and I see a lot of folks copy pasting the same solution.
Were the web action will invoke another action that does the real work in async for example using the npm package "openwhisk" and then returning 202 with the activation id back to web client.
For every call it would generate 2 action invocations and user might reach concurrent or minuteRate limits, in addition creating activation records for web actions that don't do much just schedule an action.

The proposal is to have a url for the web action that the controller will return right away with 202 and the activation id back to web client, and schedule the web action to run async.
Web Action url sync => `https://${APIHOST}/api/v1/web/guest/demo/hello`
Web Trigger url async => `https://${APIHOST}/api/v1/web/trigger/guest/demo/hello`

This might sound like a Web Hook, but I think Web Hook have a bit more complexity and a Web Hook I would say is an implementation using the Web Action sync url.
Ref: https://webhooks.pbworks.com/w/page/13385124/FrontPage

**Comment by rabbah** [comment]
Note the reason for suggesting an different URI (vs a header) is for general convenience. Often when firing a webhook configuring the http request with a custom header is not part of the available feature set from the source. 

In general calling this a trigger may suggest one can attach a rule and get 1-n broadcasting. So we’re carefully considering here that this is a 1-1 case to async invoke the webaction only for now. 

No custom codes/response or rules, yet.

**Comment by michaelmarth** [comment]
SGTM, but suggest to use /api/v1/async or /api/v1/web-async instead of /api/v1/hook
As you explained yourself, web hooks are different concepts. Even more, one can use the existing web actions as web hooks (we do that in our deployment) - so calling only the async version "hook" is confusing IMHO.

**Comment by csantanapr** [comment]
Yeah I was not sure what to use in url, went back and forth 

/web/hook/guest/...
/web/trigger/guest/...
/web/async/guest/...
/web-async/guest/...
/async/guest/...

**Comment by csantanapr** [comment]
@rabbah shoo shoo you are feature creeping 😂 

One aspect is I would to reserve /triggers/guest... for public/web triggers

This issue then is more descriptive as a plain simple Web action async
To have consistency parsing url segment I would go with one segment

With final proposal for web action async to be /web-async/guest...

To summarize:
/api/v1/web/guest/... (Today’s Web action sync blocking invoke)
/api/v1/web-async/guest/... (New Web Action Async)
/api/v1/trigger/guest/... (New Web Trigger, new issue to be created to cover this)

**Comment by rabbah** [comment]
Maybe repurpose https://github.com/apache/incubator-openwhisk/issues/1805 for the "trigger" part?

I admit I'm feature creeping - so ignore the musing for now.... there is something potentially more general here: if you're executing a sequence or composition, today the HTTP termination is tied to the result of the composition. The web-async is decoupling the HTTP termination from the activation result, and so provides a fire and forget mechanism. It is elegant because it can be done entirely in the controller.

It is restrictive on the flip side, since it doesn't however provide a way to call back/provide an id for decoupled polling (in other words, the async web action returns `Unit`, but the thought of returning a `Promise[Any]` is intriguing).

**Comment by ddragosd** [comment]
Just making sure we cover all options: was the query param `async=true` considered as well? 

On the `/trigger`: does the proposal suggest it should work with both “triggers” and “webactions”, or limit it to “webactions” only ?

**Comment by rabbah** [comment]
in a slack chat Carlos brought up the query parameter but we found that it could be ambiguous since it can be an parameter intended for the action itself.  iirc he also suggested a privileged name such as __OW_ASYNC.

**Comment by rabbah** [comment]
On /trigger my inclination is to support it only on webactions today (with an implicit trigger and rule - any webaction can be invoked in this way, asynchrinously, without creating a trigger or rule), but this will cause ambiguity later if we allow real triggers to be named so I think my suggestion was not well thought out at the time.

**Comment by csantanapr** [comment]
Yep a query and header parameter was brought up but this would need to be special named as we don’t want to conflict with a header or query intended for the action. 

On my last proposal I indicated that this issue is only to cover the minimum to do invoke an action in async way nothing to deal with triggers. 

/web/trigger can be discuss in another issue, right I see it as a way to make a trigger fire with no whisk authentication the trigger/rule/actions owner gets assigned the bill of the duration :-), allowing anything to fire the trigger.

**Comment by fmeschbe** [comment]
I don't know where this stands, but this sounds like a natural use case for the [`Prefer` request header (RFC 7240)](https://tools.ietf.org/html/rfc7240) with the [`respond-async`](https://tools.ietf.org/html/rfc7240#section-4.1) preference ...
